Gems/ROS2/Code/Source/ContactSensor/ROS2ContactSensorComponent.cpp

```diff
@@ -0,0 +1,203 @@
+        {
+            serialize->Class<ROS2ContactSensorComponent, ROS2SensorComponent>()->Version(1);
+
+            if (AZ::EditContext* editContext = serialize->GetEditContext())
+            {
+                editContext->Class<ROS2ContactSensorComponent>("ROS2 Contact Sensor", "Contact detection controller")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC_CE("Game"));
+            }
+        }
+    }
+
+    void ROS2ContactSensorComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC_CE("PhysicsColliderService"));
+        required.push_back(AZ_CRC_CE("ROS2Frame"));
+    }
+
+    void ROS2ContactSensorComponent::Activate()
+    {
+        m_entityId = GetEntityId();
+        AZ::Entity* entity = nullptr;
+        AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationRequests::FindEntity, m_entityId);
+        m_entityName = entity->GetName();
+
+        auto ros2Node = ROS2Interface::Get()->GetNode();
+        AZ_Assert(m_sensorConfiguration.m_publishersConfigurations.size() == 1, "Invalid configuration of publishers for Contact sensor");
+        const auto publisherConfig = m_sensorConfiguration.m_publishersConfigurations["gazebo_msgs::msg::ContactsState"];
+        const auto fullTopic = ROS2Names::GetNamespacedName(GetNamespace(), publisherConfig.m_topic);
+        m_contactsPublisher = ros2Node->create_publisher<gazebo_msgs::msg::ContactsState>(fullTopic.data(), publisherConfig.GetQoS());
+
+        m_onCollisionBeginHandler = AzPhysics::SimulatedBodyEvents::OnCollisionBegin::Handler(
+            [this]([[maybe_unused]] AzPhysics::SimulatedBodyHandle bodyHandle, const AzPhysics::CollisionEvent& event)
+            {
+                AddNewContact(event);
+            });
+
+        m_onCollisionPersistHandler = AzPhysics::SimulatedBodyEvents::OnCollisionPersist::Handler(
+            [this]([[maybe_unused]] AzPhysics::SimulatedBodyHandle bodyHandle, const AzPhysics::CollisionEvent& event)
+            {
+                AddNewContact(event);
+            });
+
+        m_onCollisionEndHandler = AzPhysics::SimulatedBodyEvents::OnCollisionEnd::Handler(
+            [this]([[maybe_unused]] AzPhysics::SimulatedBodyHandle bodyHandle, const AzPhysics::CollisionEvent& event)
+            {
+                AZStd::lock_guard<AZStd::mutex> lock(m_activeContactsMutex);
+                m_activeContacts.erase(event.m_body2->GetEntityId());
+            });
+
+        ROS2SensorComponent::Activate();
+    }
+
+    void ROS2ContactSensorComponent::Deactivate()
+    {
+        m_activeContacts.clear();
+        m_contactsPublisher.reset();
+        m_onCollisionBeginHandler.Disconnect();
```

Comment:
Please check if those Handlers were connected before disconnecting.

Reply:
There is no need for that, the _Disconnect()_ function already checks if the Handler is connected:
```cpp
template <typename... Params>
    void EventHandler<Params...>::Disconnect()
    {
        if (m_event)
        {
            m_event->Disconnect(*this);
        }
    }
    
    template <typename... Params>
    bool EventHandler<Params...>::IsConnected() const
    {
        return m_event != nullptr;
    }
```